edit-post/hooks/blocks/media-upload/index.js

```diff
@@ -0,0 +1,175 @@
+const slimImageObject = ( img ) => {
+	const attrSet = [ 'sizes', 'mime', 'type', 'subtype', 'id', 'url', 'alt', 'link', 'caption' ];
+	return pick( img, attrSet );
+};
+
+class MediaUpload extends Component {
+	constructor( { multiple = false, type, gallery = false, title = __( 'Select or Upload Media' ), modalClass } ) {
+		super( ...arguments );
+		this.openModal = this.openModal.bind( this );
+		this.onSelect = this.onSelect.bind( this );
+		this.onUpdate = this.onUpdate.bind( this );
+		this.onOpen = this.onOpen.bind( this );
+		this.processMediaCaption = this.processMediaCaption.bind( this );
+		const frameConfig = {
+			title,
+			button: {
+				text: __( 'Select' ),
+			},
+			multiple,
+			selection: new wp.media.model.Selection( [] ),
+		};
+		if ( !! type ) {
+			frameConfig.library = { type };
+		}
+
+		if ( gallery ) {
+			const GalleryDetailsMediaFrame = getGalleryDetailsMediaFrame();
+			this.frame = new GalleryDetailsMediaFrame( {
+				frame: 'select',
+				mimeType: type,
+				state: 'gallery',
+			} );
+			wp.media.frame = this.frame;
+		} else {
+			this.frame = wp.media( frameConfig );
+		}
+
+		if ( modalClass ) {
+			this.frame.$el.addClass( modalClass );
+		}
+
+		// When an image is selected in the media frame...
+		this.frame.on( 'select', this.onSelect );
+		this.frame.on( 'update', this.onUpdate );
+		this.frame.on( 'open', this.onOpen );
+	}
+
+	componentWillUnmount() {
+		this.frame.remove();
+	}
+
+	onUpdate( selections ) {
+		const { onSelect, multiple = false } = this.props;
+		const state = this.frame.state();
+		const selectedImages = selections || state.get( 'selection' );
+
+		if ( ! selectedImages || ! selectedImages.models.length ) {
+			return;
+		}
+
+		if ( multiple ) {
+			onSelect( selectedImages.models.map( ( model ) => this.processMediaCaption( slimImageObject( model.toJSON() ) ) ) );
+		} else {
+			onSelect( this.processMediaCaption( slimImageObject( ( selectedImages.models[ 0 ].toJSON() ) ) ) );
+		}
+	}
+
+	onSelect() {
+		const { onSelect, multiple = false } = this.props;
+		// Get media attachment details from the frame state
+		const attachment = this.frame.state().get( 'selection' ).toJSON();
+		onSelect(
+			multiple ?
+				attachment.map( this.processMediaCaption ) :
+				this.processMediaCaption( attachment[ 0 ] )
+		);
+	}
+
+	onOpen() {
+		const selection = this.frame.state().get( 'selection' );
+		const addMedia = ( id ) => {
+			const attachment = wp.media.attachment( id );
+			attachment.fetch();
+			selection.add( attachment );
+		};
+
+		if ( ! this.props.value ) {
+			return;
+		}
+
+		if ( this.props.multiple ) {
+			this.props.value.map( addMedia );
+		} else {
+			addMedia( this.props.value );
+		}
+	}
+
+	openModal() {
+		this.frame.open();
+	}
+
+	processMediaCaption( mediaObject ) {
+		return ! mediaObject.caption ?
+			mediaObject :
+			{ ...mediaObject, caption: parseWithAttributeSchema( mediaObject.caption, {
+				source: 'children',
```

Comment:
Minor: edit post is showing a deep knowledge of how blocks work by using "source: 'children'",. I understand the reason for this but I think we can also create a function in parse called "parseHtml" or something similar that just receives a string and returns the parse result (calling parseWithAttributeSchema). So each time we want to pass html we just call it, it abstracts a little bit more how parsing works.

Reply:
I'm not sure what would be the best way to go. This one is very flexible and built on top of `hpq`. I didn't want to introduce too granular function which will be used only in one place. This is definitely something to look at and a subject to further refinement.